### PR TITLE
Filter out archived repos

### PR DESCRIPTION
**Pull request by bjoernhaeuser**
I tried to figure a more way which is more aligned with the rest of the
code base, but I failed.

This seams to be the easiest fix to me.

See https://issues.jenkins-ci.org/browse/JENKINS-55019 as reference.

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 1]
```diff
@@ -955,6 +955,13 @@ public void visitSources(SCMSourceObserver observer) throws IOException, Interru
                             if (!repo.getOwnerName().equals(repoOwner)) {
                                 continue; // ignore repos in other orgs when using GHMyself
                             }
+
+                            if (repo.isArchived()) {
+                                witness.record(repo.getName(), false);
+
+                                continue;
+                            }
+
```
Comment (bitwiseman) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: Can Org repo's also be archived?

Comment (bjoernhaeuser) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: Yes :) We already did that quite often.

Comment (bitwiseman) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: @bjoernhaeuser 
Will need to put this if the "org" code path as well then?

**Comment by bitwiseman** [comment]
@bjoernhaeuser We'll need some tests for this as well.

**Comment by bjoernhaeuser** [comment]
@bitwiseman Thanks for the swift review. I put the code block in all places. 

Which kind of tests are you thinking of? I am was executing the complete test suite for the plugin and no tests is covering these methods. Could you give me a head start of writing the appropriate tests?

**Comment by bitwiseman** [comment]
@bjoernhaeuser 
I'm somewhat new to this plugin, but I can tell you what I know.

This exercises that class: https://github.com/jenkinsci/github-branch-source-plugin/blob/master/src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorTest.java

We use WireMock to mock out REST API responses.  You'll need to update the responses to include an archived repo: https://github.com/jenkinsci/github-branch-source-plugin/blob/master/src/test/resources/api/__files/body-cloudbeers-repos-v3FBW.json

You might try that on a different branch (without your change) to see adding an archived repo breaks any existing tests (probably not).  Then try some debugging here: 
https://github.com/jenkinsci/github-branch-source-plugin/blob/5b2434f39e3431b568acfcb937536b5bb3616bbc/src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorTest.java#L177-L178

That has the call to `visitSources`.  See if you can get that test to detect the archived repo.  Then apply your change and change the test to make sure the archived repo is not detected.  

You'll notice that the test currently is working against the org `cloudbeers`: 
https://github.com/jenkinsci/github-branch-source-plugin/blob/master/src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorTest.java#L143

You might also try adding a navigator for "stephenc" to the test and adding the appropriate response files to test the user path. 

This isn't the easiest system to work with, but it doesn't the job. You can use #222 as an example.  Feel free to ping me on gitter and I'm happy to help.

**Comment by bjoernhaeuser** [comment]
@bitwiseman I think I was able to add at least one test for the organization use case. Could you check if I am going down the right road? Thanks!

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 2]
```diff
@@ -104,6 +87,19 @@
 import org.kohsuke.stapler.QueryParameter;
 import org.kohsuke.stapler.interceptor.RequirePOST;
 
+import javax.inject.Inject;
```
Comment (bitwiseman) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: Revert this reordering.  We use simple alphabetical.

**Comment by bitwiseman** [comment]
@bjoernhaeuser 
Looks reasonable to me.  But I'm a little vague about what the tests are saying.

**Comment by bjoernhaeuser** [comment]
@bitwiseman The purpose of them is that they are not getting scanned. I I understand the code correctly not being scanned would mean that they are schedule for removal afterwards. Is this not correct?

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorTest.java [hunk 3]
```diff
@@ -178,6 +178,42 @@ public void addAttribute(@NonNull String key, @Nullable Object value)
...
+            }
+
+            @Override
+            public void addAttribute(@NonNull String key, @Nullable Object value)
+                throws IllegalArgumentException, ClassCastException {
+
+            }
+        };
+        navigator.visitSources(SCMSourceObserver.filter(observer, "basic", "advanced"));
+        assertThat(names, Matchers.contains("basic"));
```
Comment (bitwiseman) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorTest.java]:
```suggestion
        // basic is not archived, advanced is archived - advanced should be filtered out
        assertThat(names, Matchers.contains("basic"));
```

**Comment by bjoernhaeuser** [comment]
@bitwiseman Do we need more test coverage?

**Comment by bitwiseman** [comment]
@bjoernhaeuser 
More test coverage definitely.  You've covered one of the cases, we should cover each of them. 

I have a separate concern in that there is no way to turn this off.  I'm sure there are people who archive their repos but do not want their jobs to disappear (but do want jobs to be deleted when repo's are deleted).  Sorry to move the goal post on you a bit but this seems like a behavior that could cause bad surprises.

**Comment by joshk0** [comment]
I'm willing to help, what are the other test cases that are missing? It's not abundantly clear at first glance.

**Comment by bjoernhaeuser** [comment]
@bitwiseman 

I first tried to address the test coverage. I managed to add a test-case for the user and non organization case. Though anonymous user is missing. I can't get Wiremock to create the mapping files for me, which I then would be able to change accordingly.

Could you help me out here?

**Comment by bjoernhaeuser** [comment]
@bitwiseman do you have any input on this? :)

**Comment by joshk0** [comment]
@bitwiseman,
I would like to help with this plugin as we have many archived repos in our org. could you provide feedback to @bjoernhaeuser (i can also take a shot at addressing the feedback myself if he does not respond) so this feature can be included to the plugin?
Thanks.

**Comment by nickgriffiths** [comment]
I hope no-one minds, but I decided to have another go at this in a separate PR – #274.

**Comment by bjoernhaeuser** [comment]
Superseeded by #274

**Pull request closed without merging**